### PR TITLE
158039176 selection context

### PR DESCRIPTION
**Pull request by mjbradvica**
Initial pull request.

### CODE REVIEW

File: src/Housing.Selection.Context/DataAccess/IDbContext.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    public interface IDbContext
+    {
+        int saveChanges(); //Change to SaveChanges
+        DbSet<Room> Rooms { get; set; }
+        DbSet<Batch> Batches { get; set; }
+        DbSet<User> Users { get; set; }
+        DbSet<Address> Addresses { get; set; }
+        DbSet<Name> Names { get; set; }
+
+        //Clean up space around class and remove unnessary namespaces.
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/IDbContext.cs]: undone changes in this file. use SaveChangesAsync

File: src/Housing.Selection.Context/DataAccess/AddressRepository.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        public Address GetAddressByAddressId(Guid addressId)
+        {
+            return _housingSelectionDbContext.Addresses.First(x => x.AddressId == addressId);
+        }
+
+        public void SaveChanges()
+        {
+            _housingSelectionDbContext.saveChanges();
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/AddressRepository.cs]: should be async - likewise for other repos

File: src/Housing.Selection.Context/DataAccess/HousingSelectionDbContext.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+        }
+
+        public DbSet<User> Users { get; set; }
+        public DbSet<Room> Rooms { get; set; }
+        public DbSet<Address> Addresses { get; set; }
+        public DbSet<Batch> Batches { get; set; }
+        public DbSet<Name> Names { get; set; }
+
+        //Not needed, remove.
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/HousingSelectionDbContext.cs]: leftover changes in this code

File: src/Housing.Selection.Context/DataAccess/AddressRepository.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        public IEnumerable<Address> GetAddresses()
+        {
+            return _housingSelectionDbContext.Addresses;
+        }
+
+        public Address GetAddressById(Guid id)
+        {
+            return _housingSelectionDbContext.Addresses.First(x => x.Id == id);
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/AddressRepository.cs]: use `FirstAsync`. likewise for all the linq calls which access the db.

File: src/Housing.Selection.Testing/Context/TestAddressRepository.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        [Fact]
+        public void CanReturnAddresses()
+        {
+            var addressRepository = new AddressRepository(mockHousingContext);
+
+            var testAddresses = addressRepository.GetAddresses();
+
+            Assert.NotNull(testAddresses);
```
Comment (escalonn) [inline src/Housing.Selection.Testing/Context/TestAddressRepository.cs]: not a great test, the ienumerable will normally not be null ever. add something to the mock context and then make sure you get it back by awaiting testAddresses.ToListAsync() or SingleAsync() or something.

File: src/Housing.Selection.Context/Housing.Selection.Context.csproj [hunk 6]
```diff
@@ -1,5 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
+  <ItemGroup>
+    <Compile Remove="Interfaces\**" />
+    <EmbeddedResource Remove="Interfaces\**" />
+    <None Remove="Interfaces\**" />
+  </ItemGroup>
```
Comment (escalonn) [inline src/Housing.Selection.Context/Housing.Selection.Context.csproj]: why is this here?

File: src/Housing.Selection.Context/HttpRequests/ServiceUserCalls.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+                    if (!validate) throw new Exception("Address was not valid.");
+                }
+
+                if (user.Location == "") throw new Exception("Location was invalid.");
+
+                var response = await Client.PutAsync<ApiUser>(ApiPath.GetUserServicePath(), user);
+                if (!response.IsSuccessStatusCode)
+                {
+                    throw new Exception("Update failed for " + user.UserId);
```
Comment (escalonn) [inline src/Housing.Selection.Context/HttpRequests/ServiceUserCalls.cs]: make sure we are writing these exceptions to log somewhere (like controller) if not here

File: src/Housing.Selection.Context/Polling/IPollBatch.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Housing.Selection.Library.HousingModels;
+using Housing.Selection.Library.ServiceHubModels;
+
+namespace Housing.Selection.Context.Polling
+{
+    public interface IPollBatch
+    {
+        Task<List<Batch>> BatchPoll();
```
Comment (escalonn) [inline src/Housing.Selection.Context/Polling/IPollBatch.cs]: method should be named ending in `Async`. likewise for rest

File: src/Housing.Selection.Context/Polling/PollUser.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            return housingUser;
+        }
+
+        public Room GetRoomId(ApiUser user, List<ApiRoom> rooms)
+        {
+            var roomId = (from x in rooms
+                          where x.Address.AddressId == user.Address.AddressId
+                          select x.RoomId).FirstOrDefault();
+            if(roomId != null)
+                return roomRepository.GetRoomByRoomId(roomId);
```
Comment (escalonn) [inline src/Housing.Selection.Context/Polling/PollUser.cs]: access like this should be async

File: src/Housing.Selection.Context/Selection/AUserFilter.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
+﻿using Housing.Selection.Library.HousingModels;
+using Housing.Selection.Library.ViewModels;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Housing.Selection.Context.Selection
+{
+    public abstract class AUserFilter
```
Comment (escalonn) [inline src/Housing.Selection.Context/Selection/AUserFilter.cs]: no documentation

File: src/Housing.Selection.Context/Selection/ARoomFilter.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Housing.Selection.Library.HousingModels;
+using Housing.Selection.Library.ViewModels;
+
+namespace Housing.Selection.Context.Selection
+{
+    public abstract class ARoomFilter
```
Comment (escalonn) [inline src/Housing.Selection.Context/Selection/ARoomFilter.cs]: no documentation

File: src/Housing.Selection.Context/Polling/PollRoom.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Housing.Selection.Context.DataAccess;
+using Housing.Selection.Context.HttpRequests;
+using Housing.Selection.Library.HousingModels;
+using Housing.Selection.Library.ServiceHubModels;
+
+namespace Housing.Selection.Context.Polling
+{
+    public class PollRoom : IPollRoom
```
Comment (escalonn) [inline src/Housing.Selection.Context/Polling/PollRoom.cs]: no docs on this or similar

File: src/Housing.Selection.Library/HousingModels/Batch.cs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                BatchId = apiBatch.BatchId,
+                StartDate = (DateTime) apiBatch.StartDate,
+                EndDate = (DateTime) apiBatch.EndDate,
+                BatchName = apiBatch.BatchName,
+                BatchOccupancy = (int) apiBatch.BatchOccupancy,
+                BatchSkill = apiBatch.BatchSkill,
+                Location = apiBatch.Location
+            };
+            //TODO - Figure out how to handle apiBatch userIds
```
Comment (escalonn) [inline src/Housing.Selection.Library/HousingModels/Batch.cs]: don't forget about this

**Comment by mjbradvica** [comment]
Going to do logging for exceptions when we do the pull request with the api controller.

File: src/Housing.Selection.Context/DataAccess/AddressRepository.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            return await _housingSelectionDbContext.Addresses.FirstAsync(x => x.Id == id);
+        }
+
+        public async Task<Address> GetAddressByAddressId(Guid addressId)
+        {
+            return await _housingSelectionDbContext.Addresses.FirstAsync(x => x.AddressId == addressId);
+        }
+
+        public async Task SaveChanges()
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/AddressRepository.cs]: this method should be named SaveChangesAsync. likewise for other repos.

File: src/Housing.Selection.Context/DataAccess/IAddressRepository.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Housing.Selection.Context.DataAccess
+{
+   public interface IAddressRepository
+    {
+        IEnumerable<Address> GetAddresses();
+        Task <Address> GetAddressById(Guid id);
+        Task <Address> GetAddressByAddressId(Guid nameId);
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/IAddressRepository.cs]: formatting

File: src/Housing.Selection.Context/DataAccess/IDbContext.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public interface IDbContext
+    {
+        int SaveChanges();
+        DbSet<Room> Rooms { get; set; }
+        DbSet<Batch> Batches { get; set; }
+        DbSet<User> Users { get; set; }
+        DbSet<Address> Addresses { get; set; }
+        DbSet<Name> Names { get; set; }
+        Task<int> SaveChangesAsync( CancellationToken ct = default(CancellationToken));
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/IDbContext.cs]: formatting

File: src/Housing.Selection.Context/HttpRequests/IApiPathBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,14 @@
+using System;
+
+namespace Housing.Selection.Context.HttpRequests
+{
+    public interface IApiPathBuilder
+    {
+        string GetBatchServicePath ();
```
Comment (escalonn) [inline src/Housing.Selection.Context/HttpRequests/IApiPathBuilder.cs]: formatting

File: src/Housing.Selection.Context/Polling/IPollBatch.cs (hunk 8 above)
Comment (escalonn) [reply src/Housing.Selection.Context/Polling/IPollBatch.cs]: method is still not named async, nor for other poll classes...

File: src/Housing.Selection.Context/Polling/PollRoom.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            var roomList = new List<Room>();
+            var rooms = await _roomRetrieval.RetrieveAllRoomsAsync();
+            if (rooms != null)
+            {
+                foreach (var room in rooms)
+                {
+                    roomList.Add(await UpdateRoom(room));
+                }
+            }
+            return roomList;
```
Comment (escalonn) [inline src/Housing.Selection.Context/Polling/PollRoom.cs]: it's not a priority but just FYI we prefer to write code like this with `WhenAll`

File: src/Housing.Selection.Context/Polling/PollUser.cs [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+        /// </summary>
+        /// <returns>
+        /// Returns a Task<List<User>> that contains the updated user list
+        /// </returns>
+        public async Task<List<User>> UserPoll()
+        {
+            var userList = new List<User>();
+            var users = await _userRetrieval.RetrieveAllUsersAsync();
+            var batches = await _batchRetrieval.RetrieveAllBatchesAsync();
+            var rooms = await _roomRetrieval.RetrieveAllRoomsAsync();
```
Comment (escalonn) [inline src/Housing.Selection.Context/Polling/PollUser.cs]: this would also be a candidate for WhenAll if we had to time to focus on such things

File: src/Housing.Selection.Context/Selection/ARoomFilter.cs [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            Successor?.FilterRequest(ref filterRooms, roomSearchViewModel);
+        }
+    }
+
+    public class IsCompletelyUnassignedFilter : ARoomFilter
+    {
+        public override void FilterRequest(ref List<Room> filterRooms, RoomSearchViewModel roomSearchViewModel)
+        {
+            if(roomSearchViewModel.IsCompletelyUnassigned)
```
Comment (escalonn) [inline src/Housing.Selection.Context/Selection/ARoomFilter.cs]: formatting in this file and userfilter

File: src/Housing.Selection.Context/Selection/SelectionService.cs [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+            await _userRepository.SaveChanges();
+            await _roomRepository.SaveChanges();
+
+            var apiUser = _mapper.Map<ApiUser>(newUser);
+            var apiRoom = _mapper.Map<ApiRoom>(addRoom);
+
+            await _serviceUserCalls.UpdateUserAsync(apiUser);
+            await _serviceRoomCalls.UpdateRoomAsync(apiRoom);
+        }
```
Comment (escalonn) [inline src/Housing.Selection.Context/Selection/SelectionService.cs]: always `async Task`, never `async void`. also the name should end in `Async`.

File: src/Housing.Selection.Context/Selection/SelectionService.cs [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+        public IEnumerable<Room> CustomSearch(RoomSearchViewModel roomSearchViewModel)
+        {
+            var returnedRooms = _roomRepository.GetRooms().ToList();
```
Comment (escalonn) [inline src/Housing.Selection.Context/Selection/SelectionService.cs]: db access. `ToListAsync`. likewise for the rest of this file. all will wind up as async methods, likewise everything that calls them.

File: src/Housing.Selection.Context/ServiceHubProxies/ServiceBatchCallProxy.cs [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+            for (int i = 0; i < number; i++)
+            {
+                ids.Add(Guid.NewGuid());
+            }
+            return ids;
+        }
+     
+        public async Task<List<ApiBatch>> RetrieveAllBatchesAsync()
+        {
+            return  await Task.Run<List<ApiBatch>>(() => _batches);
```
Comment (escalonn) [inline src/Housing.Selection.Context/ServiceHubProxies/ServiceBatchCallProxy.cs]: use `return Task.FromResult(_batches)` and take off the `async` modifier from this implementation. that's what we do when synchronously implementing an async interface. likewise for other two proxies

File: src/Housing.Selection.Context/DataAccess/AddressRepository.cs [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+
+        public void AddAddress(Address address)
+        {
+            _housingSelectionDbContext.Addresses.Add(address);
+        }
+
+        public IEnumerable<Address> GetAddresses()
+        {
+            return _housingSelectionDbContext.Addresses;
+        }
```
Comment (escalonn) [inline src/Housing.Selection.Context/DataAccess/AddressRepository.cs]: all each repo method returning ienumerable like this, do "Find All References", and ensure they are all accessing it asynchronously.

File: src/Housing.Selection.Context/Selection/ARoomFilter.cs (hunk 20 above)
Comment (mjbradvica) [reply src/Housing.Selection.Context/Selection/ARoomFilter.cs]: Could you extrapolate what you mean here? Should these all be in separate files or....?

**Comment by mjbradvica** [comment]
PR updated
